### PR TITLE
add experimental zson writer

### PR DESCRIPTION
**Pull request by mccanne**
This commit adds a first cut at the zson writer.  It came together quite
nicely and is ready for people to try out and provide feedback.
In subsequent PRs, we'll add an update to the spec (issue #1678),
comprehensive tests, reader support (issue #1680), and misc other
finishes like enums and proper string escaping.

### CODE REVIEW

File: zio/zsonio/ztests/first-test.yaml [hunk 1]
```diff
@@ -0,0 +1,118 @@
+zql: "*"
```
Comment (mattnibs) [inline zio/zsonio/ztests/first-test.yaml:1]: Can we get a better more descriptive test than `first-test.yaml`? Ideally shouldn't these be broken up by test case?

File: zio/zsonio/writer.go [hunk 2]
```diff
@@ -0,0 +1,366 @@
...
+	"github.com/brimsec/zq/zng"
+	"github.com/brimsec/zq/zng/resolver"
+)
+
+type Writer struct {
+	writer      io.WriteCloser
+	zctx        *resolver.Context
+	mapper      *resolver.Mapper
+	tags        typemap
+	nrec        int
```
Comment (mattnibs) [inline zio/zsonio/writer.go:22]: This field is unused.

File: zio/zsonio/ztests/first-test.yaml [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+      _path: "conn",
+      ts: 2018-03-24T17:15:20.600725Z,
+      id: {
+          orig_h: 10.47.1.152,
+          orig_p: 49562 (uint16) (port),
+          resp_h: 23.217.103.245,
+          resp_port: 80 (port)
+      } (25) (flow),
+      proto: "tcp" (zenum)
+  } (24),
```
Comment (mattnibs) [inline zio/zsonio/ztests/first-test.yaml:38]: I'm sorry if I missed the explanation, but why do these need a typeid next to them?

File: zio/zsonio/ztests/first-test.yaml (hunk 1 above)
Comment (nwt) [reply zio/zsonio/ztests/first-test.yaml:1]: first-test.yaml is an instant classic!

File: zio/zio.go [hunk 4]
```diff
@@ -23,6 +24,7 @@ type WriterOpts struct {
 	EpochDates bool
 	Text       textio.WriterOpts
 	Zng        zngio.WriterOpts
+	ZSON       zsonio.WriterOpts
```
Comment (nwt) [inline zio/zio.go:27]: Nit: We should be consistent about capitalization of acronyms, either by changing this to Zson or (in another PR) changing Zng and Zst to ZNG and ZST.

**Review by nwt — APPROVED** [review]

File: zio/zsonio/ztests/first-test.yaml (hunk 3 above)
Comment (mccanne) [reply zio/zsonio/ztests/first-test.yaml:38]: If the body of a record is fully type-defined, then ZSON doezs not required the extra type decorator but it's also ok to have it.  The writer doesn't have logic to determine if it's redundant so it always includes it for now.  The reader (coming soon) will handle both cases correctly.

File: zio/zsonio/writer.go [hunk 5]
```diff
@@ -0,0 +1,366 @@
...
+				return err
+			}
+		}
+		return w.writeDecorator(decType)
+	}
+	var err error
+	switch t := typ.(type) {
+	//XXX Need enum support. See #1676.
+	default:
+		err = w.writePrimitive(indent, typ, bytes)
```
[the remark this answers is not in the document]
Comment (mccanne) [reply zio/zsonio/writer.go:79]: I like putting default first when it's a class of things that is operated on then the following things are the finer-grained exceptions.  When the default case is error or null, I think it's better going last.  Maybe we could discuss and agree on the convention?   I'm of course happy to go with the consensus here.  I will leave this one as is for now.

File: zio/zio.go (hunk 4 above)
Comment (mccanne) [reply zio/zio.go:27]: I would vote for the latter.

**Pull request merged**